W-14237985 mtf (#232)
mulesoft/docs-mule-sdk: modules/ROOT/pages/mtf-metadata-testing.adoc

= Metadata 
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

Test the metadata defined for your module.

== Before You Begin

> * Know how to define metadata in your module. For more information, refer to xref:metadata.adoc[].

Reviewer: I know that "Understand" is a 25-cent word and "Know" is a 5-cent word (as my previous editor used to say), but I wonder if it's more precise in this case.

> * To create metadata tests, add the `mtf-tools` dependency to your project:

Reviewer: Do you need to specify the filename here, or is it obvious from the larger context?

Author: I'd say it is obvious

+
[source,xml,linenums]
----
<dependency>
    <groupId>com.mulesoft.munit</groupId>
    <artifactId>mtf-tools</artifactId>
    <version>1.0.0</version>
    <classifier>mule-plugin</classifier>
    <scope>test</scope>
</dependency>
----

> CAUTION: Metadata testing is available since Mule runtime engine 4.2.0. When running your tests against an earlier version, suites containing metadata tests are ignored.

Reviewer: ```suggestion
CAUTION: Metadata testing is available as of Mule runtime engine version 4.2.0. When running your tests against an earlier version, suites containing metadata tests are ignored.
```

== Tooling Test

> The `tooling-test` component validates the module's metadata at design time, unlike the `munit:test` component that validates the module's metadata at execution. The `tooling-test` component is represented as follows:

Reviewer: ```suggestion
The `tooling-test` component validates the module's metadata at design time, unlike the `munit:test` component, which validates the module's metadata at execution. The following code describes the `tooling-test` component:
```

Reviewer: Do you need to mention the munit test here?

Reviewer: (Just an active voice suggestion.)

[source,xml,linenums]
----
<mtf:tooling-test name="metadataTest">
    ...
</mtf:tooling-test>
----

> When a suite with a tooling test runs, XML validations are disabled. When testing metadata, it might be necessary to avoid filling required parameters.

Reviewer: What does this second sentence mean?
You might want to avoid testing required parameters?

The tooling test can have the following parameters:

[%header%autowidth.spread,cols="a,a"]
|===
|Name |Description

|`name`

> | Mandatory. Defines the name of the test. The name must be unique inside the test suite.

Reviewer: Are required and mandatory the same?

|`description`

> |Describes the scenario that is tested.

Reviewer: ```suggestion
|Describes the scenario to test.
```

|`ignore`

> |Defines whether the test is ignored. If this parameter is not present, the test is not ignored.

Reviewer: ```suggestion
|Specifies whether to ignore the test. If this parameter is not present, the test is not ignored.
```

Reviewer: (Not sure if these active wording suggestions are correct - use your judgement)

|`tags`

> |Comma-separated list of tags to assign to the test.

Reviewer: ```suggestion
|Specifies a comma-separated list of tags to assign to the test.
```

Reviewer: For table rows (and lists), strive for consistent (such as all noun or all imperative) structure.

|`expectFailureCode`

> |Defines the failure code that is received after the metadata calculation of this test.

Reviewer: ```suggestion
|Defines the failure code that is received after calculating the metadata for this test.
```

Reviewer: That suggestion is probably not better.

Author: I like it!

|`expectFailureMessage`
|Defines the failure message that is received after the metadata calculation of this test.

|===

The tooling test has two main sections:

* Metadata Scopes
+
Wraps the component in which to calculate metadata and defines the metadata information to obtain. Possible values are:
+
** Metadata keys
** Input metadata
** Output metadata
** Attributes metadata

* Validation
+
Contains all the validations regarding the result of the metadata scope.

== Metadata Scopes

> Use metadata scopes in the tooling test. You can use only one metadata scope in each test. You can use either an operation or a source in each of these scopes.

Reviewer: Can you combine the last two sentences?

=== Metadata Keys

The `metadata keys` scope obtains the metadata keys for the module. The `payload` contains a map, in which the key is the metadata key ID and the value is the metadata key:

[source,xml,linenums]
----
<mtf:tooling-test name="keysTest">
    <mtf:get-metadata-keys>
        <example:my-operation config-ref="config" />
    </mtf:get-metadata-keys>
    <mtf:validation>
        <munit-tools:assert-equals actual="#[payload.aKeyId.displayName]" expected="#['Display Name']"/>
    </mtf:validation>
</mtf:tooling-test>
----

=== Input Metadata

The `input metadata` scope obtains the metadata type for the specified parameter of the module. The `payload` contains a JSON representation of the metadata type:

[source,xml,linenums]
----
<mtf:tooling-test name="inputMetadataTest">
    <mtf:get-input-metadata parameter="aParam">
        <example:my-operation config-ref="config"/>
    </mtf:get-input-metadata>
    <mtf:validation>
        <munit-tools:assert-equals actual="#[payload.'type']" expected="#['Number']"/>
    </mtf:validation>
</mtf:tooling-test>
----

> === Output Metadata

Reviewer: These are all nicely parallel, btw.

Reviewer: I wonder if there's a way to present that's more scannable. 🤔

> The `output metadata` obtains the metadata type for the output `payload` of the module. The `payload` contains a JSON representation of the metadata type:

Reviewer: ```suggestion
The `output metadata` obtains the metadata type for the output `payload` of the module, which contains a JSON representation of the metadata type:
```

        <munit-tools:assert-equals actual="#[payload.fields[0].key.name]" expected="#['lastName']"/>
        <munit-tools:assert-equals actual="#[payload.fields[0].model.'type']" expected="#['String']"/>
        <munit-tools:assert-equals actual="#[payload.fields[1].key.name]" expected="#['name']"/>
        <munit-tools:assert-equals actual="#[payload.fields[1].model.'type']" expected="#['String']"/>
    </mtf:validation>
</mtf:tooling-test>
----

=== Attributes Metadata

The `attributes metadata` scope obtains the metadata type for the `attributes` metadata of the module. The `payload` contains a JSON representation of the metadata type:

[source,xml,linenums]
----
<mtf:tooling-test name="attributesMetadataTest">
    <mtf:get-attributes-metadata>
        <example:my-operation config-ref="config" />
    </mtf:get-attributes-metadata>
    <mtf:validation>
        <munit-tools:assert-equals actual="#[payload.format]" expected="#['java']"/>
        <munit-tools:assert-equals actual="#[payload.fields[0].key.name]" expected="#['age']"/>
        <munit-tools:assert-equals actual="#[payload.fields[0].model.'type']" expected="#['Number']"/>
    </mtf:validation>
</mtf:tooling-test>
----

> The `mtf:get-values` element retrieves the possible values for the specified parameter of the module. The `payload` contains a JSON representation of the values.

Reviewer: Does this need a heading or is it part of the previous example?

Author: Part of the previous example

[source,xml,linenums]
----
<mtf:tooling-test name="attributesMetadataTest">
    <mtf:get-values parameter="aParam">
        <example:my-operation config-ref="config"/>
    </mtf:get-values>
    <mtf:validation>
        <munit-tools:assert-equals actual="#[payload.WRITER.displayName]" expected="WRITER"/>
        <munit-tools:assert-equals actual="#[payload.READER.displayName]" expected="READER"/>
        <munit-tools:assert-equals actual="#[payload.ADMIN.displayName]" expected="ADMIN"/>
    </mtf:validation>
</mtf:tooling-test>
----

== Validation

> A common use case when testing metadata is validating failures that might occur during metadata calculation. To validate these failures, use the `expectFailureCode` and `expectFailureMessage` parameters:

Reviewer: Nice short description!

[source,xml,linenums]
----
<mtf:tooling-test name="expectFailureTest"
        expectFailureCode="INVALID_METADATA_KEY"
        expectFailureMessage="No metadata available for an empty name">
    <mtf:get-output-metadata>
        <example:my-operation config-ref="config" id=""/>
    </mtf:get-output-metadata>
</mtf:tooling-test>
----

=== Assert Type Processor

Use the assert type processor to assert the metadata type structure returned by metadata scopes such as `get-input-metadata`, `get-output-metadata`, and `get-attributes-metadata`.

> The assert type processor compares the structure of both metadata types and ignores fields such as `format` or `annotations`. To perform assertions over these fields, you can use the `payload` result of the metadata scopes operations as shown in the previous examples.

Reviewer: ```suggestion
The assert type processor compares the structure of both metadata types and ignores fields such as `format` or `annotations`. To perform assertions over these fields, use the `payload` result of the metadata scopes operations as shown in the previous examples.
```

If the assertion fails, the processor throws a `java.lang.AssertionError` error.

> ==== From Class

Reviewer: This is getting deep, hierarchically. Maybe OK?

Reviewer: Maybe a DL for these H4s?

Author: DL = description list? 

It is getting hierarchically deep, but I think keeping it as a separate header gives more visibility. Also if it were in a list, it might be messy since each bullet point would have a code snippet under it. What do you think?

Reviewer: I just looked at it in context and you're right - there's enough stuff in each H4 to justify a section.

To compare a metadata type structure with another metadata type structure obtained from a Java class, use the `fromClass` parameter:

[source,xml,linenums]
----
<mtf:tooling-test name="assertTypeFromClass">
    <mtf:get-output-metadata>
        <example:my-operation config-ref="config" />
    </mtf:get-output-metadata>
    <mtf:validation>
        <mtf:assert-type actual="#[payload]" fromClass="com.example.Person"/>
    </mtf:validation>
</mtf:tooling-test>
----

The `actual` field defaults to `payload`. You can choose to not specify the `actual` field. For example:

[source,xml,linenums]
--
<mtf:assert-type fromClass="com.example.Person"/>
--

==== From Schema

To compare a metadata type structure with another metadata type structure obtained from a schema, use the `fromSchema` parameter.

> The `fromSchema` parameter supports metadata types such as JSON schema, XSD schema, and RAML data types. The file extensions must be `.json`, `.xsd`, and `.raml` respectively and must be stored in the `src/test/resources` folder.

Reviewer: ```suggestion
The `fromSchema` parameter supports metadata types, such as JSON schema, XSD schema, and RAML data types. The file extensions must be `.json`, `.xsd`, and `.raml` respectively, and must be stored in the `src/test/resources` folder.
```

Reviewer: Are these the only metadata types? If so, remove the "such as".

Reviewer: If others, then... what are the file extensions required?

Author: Like this?

The `fromSchema` parameter supports metadata types JSON schema, XSD schema, and RAML data types. The file extensions must be `.json`, `.xsd`, and `.raml` respectively, and must be stored in the `src/test/resources` folder.

Reviewer: Perfect

    </mtf:get-output-metadata>
    <mtf:validation>
        <munit-tools:assert-equals actual="#[payload.format]" expected="#['java']"/>
        <munit-tools:assert-equals actual="#[payload.fields[0].key.name]" expected="#['name']"/>
        <munit-tools:assert-equals actual="#[payload.fields[0].model.'type']" expected="#['String']"/>
        <munit-tools:assert-equals actual="#[payload.fields[1].key.name]" expected="#['age']"/>
        <munit-tools:assert-equals actual="#[payload.fields[1].model.'type']" expected="#['Number']"/>
    </mtf:validation>
</mtf:tooling-test>
----

== Test an Expected Failure

In the previous example, the module throws a `MetadataResolvingException` error when the parameter is null or empty.

The following test validates that the proper failure code and message are thrown when the parameter is invalid:

[source, xml, linenums]
----
<mtf:tooling-test name="failureTest"
                  expectFailureMessage="No metadata available for an empty param"
                  expectFailureCode="INVALID_METADATA_KEY">
    <mtf:get-output-metadata>
        <test-connector:with-metadata param=""/>
    </mtf:get-output-metadata>
</mtf:tooling-test>
----

> 

Reviewer: No See Alsos?

Author: I can add See Alsos. Is it best practice to include See Alsos on every page? Don't think I saw guidance on this

Reviewer: No. We typically only add them if there's relevant info to link to. Not a requirement, but might be a nice-to-have?
